utils/structs/vector.go

```diff
@@ -0,0 +1,282 @@
+		}
+
+		n += inc
+
+		if cap(*v) < size {
+			*v = make([]T, size)
+		}
+
+		*v = (*v)[:size]
+
+		var t T
+		switch any(t).(type) {
+		case uint, uint64, int, int64, float64:
+
+			if inc, err = buffer.ReadAsUint64Slice[T](r, *v); err != nil {
+				return n + inc, fmt.Errorf("buffer.ReadAsUint64Slice[%T]: %w", t, err)
+			}
+
+			n += inc
+
+		case uint32, int32, float32:
+
+			if inc, err = buffer.ReadAsUint32Slice[T](r, *v); err != nil {
+				return n + inc, fmt.Errorf("buffer.ReadAsUint32Slice[%T]: %w", t, err)
+			}
+
+			n += inc
+
+		case uint16, int16:
+
+			if inc, err = buffer.ReadAsUint16Slice[T](r, *v); err != nil {
+				return n + inc, fmt.Errorf("buffer.ReadAsUint16Slice[%T]: %w", t, err)
+			}
+
+			n += inc
+
+		case uint8, int8:
+
+			if inc, err = buffer.ReadAsUint8Slice[T](r, *v); err != nil {
+				return n + inc, fmt.Errorf("buffer.ReadAsUint8Slice[%T]: %w", t, err)
+			}
+
+			n += inc
+		default:
+
+			if _, isReadable := any(new(T)).(io.ReaderFrom); !isReadable {
+				return 0, fmt.Errorf("vector component of type %T does not comply to %T", t, new(io.ReaderFrom))
+			}
+
+			for i := range *v {
+				if inc, err = any(&(*v)[i]).(io.ReaderFrom).ReadFrom(r); err != nil {
+					var t T
+					return n + inc, fmt.Errorf("%T.ReadFrom: %w", t, err)
+				}
+				n += inc
+			}
+		}
+
+		return n, nil
+
+	default:
+		return v.ReadFrom(bufio.NewReader(r))
+	}
+}
+
+// MarshalBinary encodes the object into a binary form on a newly allocated slice of bytes.
+// If T is a struct, this method requires that T implements io.WriterTo.
+func (v Vector[T]) MarshalBinary() (p []byte, err error) {
+	buf := buffer.NewBufferSize(v.BinarySize())
+	_, err = v.WriteTo(buf)
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes a slice of bytes generated by
+// MarshalBinary or WriteTo on the object.
+// If T is a struct, this method requires that T implements io.ReaderFrom.
+func (v *Vector[T]) UnmarshalBinary(p []byte) (err error) {
+	_, err = v.ReadFrom(buffer.NewBuffer(p))
+	return
+}
+
+// Equal performs a deep equal.
+// If T is a struct, this method requires that T implements Equatable.
+func (v Vector[T]) Equal(other Vector[T]) (isEqual bool) {
+
+	var t T
+	switch any(t).(type) {
+	case uint, uint64, int, int64, float64:
+		return buffer.EqualAsUint64Slice([]T(v), []T(other))
+	case uint32, int32, float32:
+		return buffer.EqualAsUint32Slice([]T(v), []T(other))
+	case uint16, int16:
+		return buffer.EqualAsUint16Slice([]T(v), []T(other))
+	case uint8, int8:
+		return buffer.EqualAsUint8Slice([]T(v), []T(other))
+	default:
+
+		if _, isEquatable := any(t).(Equatable[T]); !isEquatable {
+			panic(fmt.Errorf("vector component of type %T does not comply to %T", t, new(Equatable[T])))
+		}
+
+		isEqual := true
+		for i, v := range v {
+			/* #nosec G601 -- Implicit memory aliasing in for loop acknowledged */
+			isEqual = isEqual && any(&v).(Equatable[T]).Equal(&other[i])
+		}
+
+		return isEqual
```

Comment:
Could return earlier?
```suggestion
		for i, v := range v {
			/* #nosec G601 -- Implicit memory aliasing in for loop acknowledged */
			if !any(&v).(Equatable[T]).Equal(&other[i]) {
				return false
			}
		}
		return true
```